src/dawn/Optimizer/PassSetNonTempCaches.cpp

```diff
@@ -0,0 +1,325 @@
+        if(checkReadBeforeWrite(field.AccessID))
+          cachedreadwrites--;
+        // We need one more write to flush the cache back to the variable
+        if(!checkReadOnlyAccess(field.AccessID))
+          cachedreadwrites--;
+
+        if(cachedreadwrites > 0)
+          accessIDtoDataLocality_.emplace(field.AccessID, cachedreadwrites);
+      }
+    }
+
+    for(auto& performanceMesure : accessIDtoDataLocality_) {
+      sortedAccesses_.emplace_back(performanceMesure.first, performanceMesure.second);
+      std::sort(sortedAccesses_.begin(), sortedAccesses_.end());
+    }
+  }
+
+  /// @brief for each chached variable, check if we need a filler-stage (if we find a
+  /// read-before-write) and add it if necessary, renames all occurances and sets the cache for the
+  /// temporary variable newly created
+  ///
+  /// We always fill to the extent of the given multistage and we only add one stage to fill all the
+  /// variables to reduce synchronisation overhead
+  void addFillerStages() {
+    for(int i = 0; i < std::min((int)sortedAccesses_.size(), config_.SMemMaxFields); ++i) {
+      int oldID = sortedAccesses_[i].accessID;
+
+      // Create new temporary field and register in the instantiation
+      int newID = instantiation_->nextUID();
+
+      instantiation_->setAccessIDNamePairOfField(newID, dawn::format("__tmp_cache_%i", i), true);
+
+      // Rename all the fields in this multistage
+      multiStagePrt_->renameAllOccurrences(oldID, newID);
+
+      oldIDtoNewID.emplace(oldID, newID);
+      Cache& cache = multiStagePrt_->setCache(Cache::IJ, Cache::local, newID);
+      originalNameToCache_.emplace_back(instantiation_->getOriginalNameFromAccessID(oldID), cache,
+                                        accessIDtoDataLocality_.find(oldID)->second);
+    }
+
+    // Create the cache-filler stage
+    std::vector<int> oldIDs;
+    std::vector<int> newIDs;
+    for(auto& idPair : oldIDtoNewID) {
+      bool hasReadBeforeWrite = checkReadBeforeWrite(idPair.second);
+      if(hasReadBeforeWrite) {
+        oldIDs.push_back(idPair.first);
+        newIDs.push_back(idPair.second);
+      }
+    }
+    if(oldIDs.size() > 0) {
+      auto stageBegin = multiStagePrt_->getStages().begin();
+
+      std::shared_ptr<Stage> cacheFillStage =
+          createAssignmentStage(multiStagePrt_->getEnclosingInterval(), newIDs, oldIDs);
+      multiStagePrt_->getStages().insert(stageBegin, std::move(cacheFillStage));
+    }
+  }
+
+  /// @brief we create one stage at the end of the multistage that flushes back all the variables
+  /// that we cached in temporaries
+  void addFlushStages() {
+    std::vector<int> oldIDs;
+    std::vector<int> newIDs;
+    for(auto& idPair : oldIDtoNewID) {
+      if(!checkReadOnlyAccess(idPair.second)) {
+        oldIDs.push_back(idPair.first);
+        newIDs.push_back(idPair.second);
+      }
+    }
+
+    if(oldIDs.size() > 0) {
```

Comment:
How about `if(!oldIDs.empty())`. I think this a relays the intent slightly better.